crates/kitsune_p2p/timestamp/src/lib.rs

```diff
@@ -315,28 +277,28 @@ impl Timestamp {
     ///             Some(Timestamp( 555, 499_999_998 )));
     /// ```
     pub fn checked_sub(&self, rhs: &core::time::Duration) -> Option<Timestamp> {
-        let dur_seconds: i64 = try_opt!(i64::try_from(rhs.as_secs()).ok());
-        let dur_nanos: i64 = i64::from(rhs.subsec_nanos());
-        let seconds: i64 = try_opt!(self.0.checked_sub(dur_seconds));
-        let nanos: i64 = try_opt!(i64::from(self.1).checked_sub(dur_nanos));
-        Some(try_opt!(Timestamp::normalize(seconds, nanos)))
+        let micros = rhs.as_micros();
+        if micros <= i64::MAX as u128 {
+            Some(Self(self.0.checked_sub(micros as i64)?))
+        } else {
+            None
+        }
     }
 
-    /// Convert this timestamp to fit into a sqlite integer which is
-    /// an i64. The value will be clamped between 0 and i64::MAX.
-    pub fn to_sql_micros_lossy(self) -> i64 {
-        use std::time::Duration;
-        let s = Duration::from_secs(self.0.max(0) as u64);
-        let ns = Duration::from_nanos(self.1 as u64);
-        let ts = s.checked_add(ns).unwrap_or(s);
-        ts.as_micros().clamp(0, i64::MAX as u128) as i64
+    /// Add a duration, clamping to MAX if overflow
+    pub fn saturating_add(&self, rhs: &core::time::Duration) -> Timestamp {
+        self.checked_add(rhs).unwrap_or(Self::MAX)
     }
 
-    /// Construct from microseconds
-    pub fn from_micros(micros: i64) -> Self {
-        let secs = micros / 1_000_000;
-        let nanos = (micros % 1_000_000) as u32 * 1_000;
-        Self(secs, nanos)
+    /// Subtract a duration, clamping to MIN if overflow
+    pub fn saturating_sub(&self, rhs: &core::time::Duration) -> Timestamp {
+        self.checked_sub(rhs).unwrap_or(Self::MIN)
+    }
+
+    /// Convert this timestamp to fit into a SQLite integer which is an i64.
+    /// The value will be clamped to the valid range supported by SQLite
+    pub fn into_sql_lossy(self) -> Self {
+        Self(i64::clamp(self.0, -62167219200 * MM, 106751991167 * MM))
```

Comment:
something's not right with this comment / logic... obviously sqlite can store the full range of i64 values.

Reply:
@neonphog but I believe that SQL's datetime-related functions only operate on i64s within this range.